WASM

use `target=wasm32-unknown-emscripten`

試したが失敗
bindingを生成するときのclangがmemory.hを見つけられなかった

```
emscripten-1  |   LibRaw/libraw/libraw.h:37:10: fatal error: 'memory.h' file not found
emscripten-1  |   thread 'main' panicked at crates/libraw-sys/build.rs:221:10:
emscripten-1  |   Unable to generate bindings: ClangDiagnostic("LibRaw/libraw/libraw.h:37:10: fatal error: 'memory.h' file not found\n")
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -24,3 +24,25 @@ jobs:
       - run: cargo test
       - if: matrix.os == 'ubuntu-latest'
         run: cargo clippy && cargo fmt --check
+
+  build-wasm:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
+      - uses: actions/checkout@v4
+        with:
+          repository: emscripten-core/emsdk
+          path: emsdk
+      - run: |
+          cd emsdk
+          ./emsdk install latest
+          ./emsdk activate latest
+          source ./emsdk_env.sh
+      - uses: ./.github/actions/setup
+      - run: rustup target add wasm32-unknown-emscripten
+      - run: cargo build --target wasm32-unknown-emscripten
+        env:
+          BINDGEN_EXTRA_CLANG_ARGS: --sysroot=${{github.workspace}}/emsdk/upstream/emscripten/cache/sysroot

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -35,7 +35,7 @@ jobs:
       - uses: mymindstorm/setup-emsdk@v14
       - uses: ./.github/actions/setup
       - run: which emcc
-      - run: ls $(which emcc)
+      - run: ls $(dirname $(which emcc))
       - run: rustup target add wasm32-unknown-emscripten
       - run: cargo build --target wasm32-unknown-emscripten
         env:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -34,6 +34,8 @@ jobs:
           submodules: recursive
       - uses: mymindstorm/setup-emsdk@v14
       - uses: ./.github/actions/setup
+      - run: which emcc
+      - run: ls $(which emcc)
       - run: rustup target add wasm32-unknown-emscripten
       - run: cargo build --target wasm32-unknown-emscripten
         env:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -34,8 +34,9 @@ jobs:
           submodules: recursive
       - uses: mymindstorm/setup-emsdk@v14
       - uses: ./.github/actions/setup
-      - run: which emcc
-      - run: ls $(dirname $(which emcc))
+
+      - run: ls -l $(dirname $(which emcc))/..
+      - run: ls -l $(dirname $(which emcc))/cache
       - run: rustup target add wasm32-unknown-emscripten
       - run: cargo build --target wasm32-unknown-emscripten
         env:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -34,10 +34,6 @@ jobs:
           submodules: recursive
       - uses: mymindstorm/setup-emsdk@v14
       - uses: ./.github/actions/setup
-
-      - run: ls -l $(dirname $(which emcc))/..
-      - run: ls -l $(dirname $(which emcc))/cache
+      - run: echo "BINDGEN_EXTRA_CLANG_ARGS=--sysroot=$(dirname $(which emcc))/cache/sysroot" >> $GITHUB_ENV
       - run: rustup target add wasm32-unknown-emscripten
       - run: cargo build --target wasm32-unknown-emscripten
-        env:
-          BINDGEN_EXTRA_CLANG_ARGS: --sysroot=${{github.workspace}}/emsdk/upstream/emscripten/cache/sysroot

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -32,15 +32,7 @@ jobs:
       - uses: actions/checkout@v4
         with:
           submodules: recursive
-      - uses: actions/checkout@v4
-        with:
-          repository: emscripten-core/emsdk
-          path: emsdk
-      - run: |
-          cd emsdk
-          ./emsdk install latest
-          ./emsdk activate latest
-          source ./emsdk_env.sh
+      - uses: mymindstorm/setup-emsdk@v14
       - uses: ./.github/actions/setup
       - run: rustup target add wasm32-unknown-emscripten
       - run: cargo build --target wasm32-unknown-emscripten

--- a/crates/libraw-rs/src/colordata.rs
+++ b/crates/libraw-rs/src/colordata.rs
@@ -13,9 +13,9 @@ impl ColorData {
     impl_property!(black, u32);
     impl_property!(data_maximum, u32);
     impl_property!(maximum, u32);
-    #[cfg(windows)]
+    #[cfg(any(windows, target_family = "wasm"))]
     impl_property!(linear_max, [i32; 4]);
-    #[cfg(not(windows))]
+    #[cfg(not(any(windows, target_family = "wasm")))]
     impl_property!(linear_max, [i64; 4]);
     impl_property!(fmaximum, f32);
     impl_property!(fnorm, f32);

--- a/crates/libraw-rs/src/lib.rs
+++ b/crates/libraw-rs/src/lib.rs
@@ -16,12 +16,9 @@ use imgother::ImgOther;
 use iparams::IParams;
 use lensinfo::LensInfo;
 use makernotes::Makernotes;
-use result::{handle_error, Error, Result};
+use result::{handle_error, Result};
 use sizes::Sizes;
-use std::{
-    ffi::{CStr, CString},
-    path::Path,
-};
+use std::ffi::CStr;
 use thumbnail::Thumbnail;
 
 #[derive(Debug)]
@@ -61,11 +58,12 @@ pub fn camera_list() -> Vec<String> {
 }
 
 impl LibRaw {
-    pub fn open_file<P: AsRef<Path>>(path: &P) -> Result<Self> {
+    #[cfg(not(target_family = "wasm"))]
+    pub fn open_file<P: AsRef<std::path::Path>>(path: &P) -> Result<Self> {
         let inner = unsafe { libraw_sys::libraw_init(0) };
         let path = path.as_ref();
-        let path = path.to_str().ok_or(Error::InvalidPath)?;
-        let path = CString::new(path).map_err(|_| Error::InvalidPath)?;
+        let path = path.to_str().ok_or(result::Error::InvalidPath)?;
+        let path = std::ffi::CString::new(path).map_err(|_| result::Error::InvalidPath)?;
         handle_error(unsafe { libraw_sys::libraw_open_file(inner, path.as_ptr()) })?;
         Ok(LibRaw { inner })
     }
@@ -90,17 +88,19 @@ impl LibRaw {
         handle_error(unsafe { libraw_sys::libraw_dcraw_process(self.inner) })
     }
 
-    pub fn dcraw_ppm_tiff_writer<P: AsRef<Path>>(&self, path: &P) -> Result<()> {
+    #[cfg(not(target_family = "wasm"))]
+    pub fn dcraw_ppm_tiff_writer<P: AsRef<std::path::Path>>(&self, path: &P) -> Result<()> {
         let path = path.as_ref();
-        let path = path.to_str().ok_or(Error::InvalidPath)?;
-        let path = CString::new(path).map_err(|_| Error::InvalidPath)?;
+        let path = path.to_str().ok_or(result::Error::InvalidPath)?;
+        let path = std::ffi::CString::new(path).map_err(|_| result::Error::InvalidPath)?;
         handle_error(unsafe { libraw_sys::libraw_dcraw_ppm_tiff_writer(self.inner, path.as_ptr()) })
     }
 
-    pub fn dcraw_thumb_writer<P: AsRef<Path>>(&self, path: &P) -> Result<()> {
+    #[cfg(not(target_family = "wasm"))]
+    pub fn dcraw_thumb_writer<P: AsRef<std::path::Path>>(&self, path: &P) -> Result<()> {
         let path = path.as_ref();
-        let path = path.to_str().ok_or(Error::InvalidPath)?;
-        let path = CString::new(path).map_err(|_| Error::InvalidPath)?;
+        let path = path.to_str().ok_or(result::Error::InvalidPath)?;
+        let path = std::ffi::CString::new(path).map_err(|_| result::Error::InvalidPath)?;
         handle_error(unsafe { libraw_sys::libraw_dcraw_thumb_writer(self.inner, path.as_ptr()) })
     }
 

--- a/crates/libraw-sys/build.rs
+++ b/crates/libraw-sys/build.rs
@@ -44,8 +44,9 @@ fn build(out_dir: &Path) {
 }
 
 fn bindings(out_dir: &Path) {
-    let bindings = bindgen::Builder::default()
+    let builder = bindgen::Builder::default()
         .header("LibRaw/libraw/libraw.h")
+        .clang_arg("-fvisibility=default")
         .use_core()
         .ctypes_prefix("libc")
         .generate_comments(true)
@@ -143,11 +144,11 @@ fn bindings(out_dir: &Path) {
         .blocklist_function("truncl")
         .blocklist_function("y0l")
         .blocklist_function("y1l")
-        .blocklist_function("ynl")
-        .generate()
-        .expect("Unable to generate bindings");
+        .blocklist_function("ynl");
 
-    bindings
+    builder
+        .generate()
+        .expect("Unable to generate bindings")
         .write_to_file(out_dir.join("bindings.rs"))
         .expect("Couldn't write bindings!");
 }

--- a/dockerfiles/wasm/Dockerfile
+++ b/dockerfiles/wasm/Dockerfile
@@ -1,0 +1,12 @@
+FROM emscripten/emsdk:4.0.1-arm64
+
+ENV BINDGEN_EXTRA_CLANG_ARGS="--sysroot=/emsdk/upstream/emscripten/cache/sysroot"
+ENV RUSTUP_HOME=/usr/local/rustup \
+    CARGO_HOME=/usr/local/cargo \
+    PATH=/usr/local/cargo/bin:$PATH
+RUN apt update && apt install -y clang
+
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y \
+&& rustup target add wasm32-unknown-emscripten
+
+COPY ./Cargo.toml ./Cargo.lock /app/

--- a/dockerfiles/wasm/compose.yaml
+++ b/dockerfiles/wasm/compose.yaml
@@ -1,0 +1,14 @@
+services:
+  build:
+    build:
+      context: ../..
+      dockerfile: ./dockerfiles/wasm/Dockerfile
+    volumes:
+      - type: bind
+        source: ../../target
+        target: /app/target
+      - type: bind
+        source: ../../crates
+        target: /app/crates
+    working_dir: /app
+    command: ["cargo", "build", "--release", "--target=wasm32-unknown-emscripten"]

